src/plugins/footnotes.ts

```diff
@@ -0,0 +1,249 @@
+    ids.set(container.node.attrs.id, elementID)
+  })
+  return {
+    footnotesElements: states,
+    footnotesElementIDs: ids,
+  }
+}
+
+const buildFootnotesElementState = (
+  container: [ManuscriptNode, number],
+  element: [FootnotesElementNode, number]
+): FootnotesElementState => {
+  const fn: FootnotesElementState = {
+    element,
+    inlineFootnotes: [],
+    unusedFootnoteIDs: new Set(),
+    footnotes: [],
+    labels: new Map(),
+  }
+
+  const footnotes = findChildrenByType(element[0], schema.nodes.footnote)
+  const footnoteIDs = new Set()
+  footnotes.forEach(({ node, pos }) => {
+    fn.footnotes.push([node as FootnoteNode, element[1] + pos + 1])
+    fn.unusedFootnoteIDs.add(node.attrs.id)
+    footnoteIDs.add(node.attrs.id)
+  })
+
+  let index = 0
+  const inlineFootnotes = findChildrenByType(
+    container[0],
+    schema.nodes.inline_footnote
+  )
+  inlineFootnotes.sort((a, b) => a.pos - b.pos)
+  const orderedFootnoteIDs: string[] = []
+  inlineFootnotes.forEach(({ node, pos }) => {
+    const inlineFootnote = node as InlineFootnoteNode
+    const rids = inlineFootnote.attrs.rids
+    if (rids.some((rid) => !footnoteIDs.has(rid))) {
+      return
+    }
+    if (container[1]) {
+      pos += container[1] + 1
+    }
+    fn.inlineFootnotes.push([node as InlineFootnoteNode, pos])
+    rids.forEach((rid) => {
+      if (orderedFootnoteIDs.indexOf(rid) < 0) {
+        const label = container[1]
+          ? String(++index)
+          : generateAlphaFootnoteLabel(index++)
+        fn.labels.set(rid, label)
+        fn.unusedFootnoteIDs.delete(rid)
+        orderedFootnoteIDs.push(rid)
+      }
+    })
+  })
+  fn.unusedFootnoteIDs.forEach((id) => orderedFootnoteIDs.push(id))
+  fn.footnotes.sort(
+    ([a], [b]) =>
+      orderedFootnoteIDs.indexOf(a.attrs.id) -
+      orderedFootnoteIDs.indexOf(b.attrs.id)
+  )
+
+  return fn
+}
+
+const getVersion = (
+  $new: FootnotesElementState,
+  $old?: VersionedFootnotesElementState
+) => {
+  const nids = $new.footnotes.map(([node]) => node.attrs.id)
+  const oids = $old?.footnotes.map(([node]) => node.attrs.id)
+  if (
+    isEqual(nids, oids) &&
+    isEqual($new.unusedFootnoteIDs, $old?.unusedFootnoteIDs)
+  ) {
+    return $old?.version
+  }
+  return String(version++)
+}
+
+export default (props: EditorProps) => {
+  return new Plugin<FootnotesPluginState>({
+    key: footnotesKey,
+    state: {
+      init(config, instance): FootnotesPluginState {
+        return buildPluginState(instance.doc)
+      },
+      apply(tr, value, oldState, newState): FootnotesPluginState {
+        const $old = footnotesKey.getState(oldState)
+        if (!tr.docChanged && $old) {
+          return $old
+        }
+        return buildPluginState(newState.doc, $old)
+      },
+    },
+    appendTransaction(transactions, oldState, newState) {
+      const $old = footnotesKey.getState(oldState)
+      const $new = footnotesKey.getState(newState) as FootnotesPluginState
```

Comment:
why $new needs typecasting and $old doesnt?

Reply:
Removed the cast.